lib/src/clojure_lsp/feature/file_management.clj

```diff
@@ -138,51 +140,58 @@
       new-text)))
 
-(defn analyze-changes [{:keys [uri text version]} {:keys [producer db] :as components}]
-  (loop [state-db @db]
-    (when (>= version (get-in state-db [:documents uri :v] -1))
-      (when-let [kondo-result (shared/logging-time
-                                (str "changes analyzed by clj-kondo took %s secs")
-                                (lsp.kondo/run-kondo-on-text! text uri db))]
-        (let [filename (shared/uri->filename uri)
-              old-local-analysis (get-in @db [:analysis filename])]
-          (if (compare-and-set! db state-db (-> state-db
-                                                (update-analysis uri (:analysis kondo-result))
-                                                (update-findings uri (:findings kondo-result))
-                                                (update :processing-changes disj uri)
-                                                (assoc :kondo-config (:config kondo-result))))
-            (do
-              (f.diagnostic/sync-lint-file! uri db)
-              (when (settings/get db [:notify-references-on-file-change] true)
-                (notify-references filename old-local-analysis (get-in @db [:analysis filename]) components))
-              (clojure-producer/refresh-test-tree producer [uri]))
-            (recur @db)))))))
+(defn analyze-changes [{:keys [uri text version done]} {:keys [producer db] :as components}]
+  (shared/logging-task
+    :analyze-file
+    (loop [state-db @db]
+      (when (>= version (get-in state-db [:documents uri :v] -1))
+        (when-let [kondo-result (shared/logging-time
+                                  (str "changes analyzed by clj-kondo took %s")
+                                  (lsp.kondo/run-kondo-on-text! text uri db))]
+          (let [filename (shared/uri->filename uri)
+                old-local-analysis (get-in @db [:analysis filename])]
+            (if (compare-and-set! db state-db (-> state-db
+                                                  (update-analysis uri (:analysis kondo-result))
+                                                  (update-findings uri (:findings kondo-result))
+                                                  (update :processing-changes dissoc uri)
+                                                  (assoc :kondo-config (:config kondo-result))))
+              (do
+                (deliver done :done)
+                (f.diagnostic/sync-lint-file! uri db)
+                (when (settings/get db [:notify-references-on-file-change] true)
+                  (notify-references filename old-local-analysis (get-in @db [:analysis filename]) components))
+                (clojure-producer/refresh-test-tree producer [uri]))
+              (recur @db))))))))
 
 (defn did-change [uri changes version db]
   (let [old-text (get-in @db [:documents uri :text])
-        final-text (reduce handle-change old-text changes)]
+        final-text (reduce handle-change old-text changes)
+        done (promise)]
```

Comment:
promise is a good fit for this code indeed

Reply:
Reverted, best to ignore.